### DISCUSSION TITLE
Reverse Bits 이 정말...이지모드가 맞나요 :(

### DISCUSSION THREAD

**Discussion by Zioq** [DaleStudy/leetcode-study#779] [도움요청]
https://leetcode.com/problems/reverse-bits/description/

문제를 풀려고 한시간 이상 고민하다가 해결법이 생각나지 않아, 그냥 Solution을 보고 이해한 상태에서 문제를 제출해 보려 했는데요...

설명 유튜브도 잘 이해가 가지 않아 결국은 ChatGPT의 설명까지 필요하게 되었슴니다.

LSB니 >>>니 여러가지 테크니컬 기술들이 서술되었는데,

문득 이 문제에서 물어보고 싶은것이 무엇일까? 개발자에게 물어보고 싶었던 테크니컬은 무엇이었을까 하는 원초적인 고민이 들더라구요.

(코드와 문제풀이는 이해하면 된다지만) 과연 이 문제에서 물어보고 싶었던 원초적인 문제의 핵심은 무엇이었을까요? :)

**Comment by HC-kang** [comment]
고생하셨습니다!
제 개인적인 생각으로는 사실, 이 문제는 JS로 풀라고 만든 문제가 아니라고 생각해요.. ㅋㅋ

오히려 `C`나 `Cpp`에서는 아래처럼 간단하게 정리되는 문제라서 `Easy` 단계인것 같고, JS에서는 32비트 정수라는 로우한 변수를 실질적으로 다룰 일이 없으니 더 낯설기도 하고, `>>`나 `>>>`같은, 부호를 관리하기 위한 부차적인(?) 지식까지 있어야 하니까 영 맞지 않는 것 같아요.

```c
uint32_t reverseBits(uint32_t n) {
    uint32_t result = 0;

    for (int i = 0; i < 32; i++) {
        result = (result << 1) | (n & 1);
        n = n >> 1;
    }
    return result;
}
```

결과적으로 문제에서 의도한 것 자체는 CS의 기본이라고 할 수 있는 간단한 비트연산에 대한 내용이지만, 이게 JS라는 언어(64비트 `number`밖에 모르는 바보,,)에서 다뤄지면서 발생한 해프닝이 아닐까 싶습니다.
그냥 '이런것도 있구나' 하시고, 너무 마음쓰지 않으셔도 좋을것 같아요!

**Comment by bus710** [comment]
비트패킹이 은근히 깔끔하고 성능좋게 짜기 어려운 것 같습니다. 특히 인터프리터 언어들이 더 그런면이 있는 것 같은데, 고수준 언어들 중에선 그나마 얼랭/엘릭서의 비트패킹 문법과 성능이 좋았던 것 같습니다.

**Comment by DaleSeo** [comment]
> Reverse Bits 이 정말...이지모드가 맞나요 :(

Blind75에 비트 조작 문제가 총 3문제 있는데 모두 Easy 난이도로 설정이 되어 있습니다. 하지만 실제로 풀어보면 체감되는 난이도는 약간씩 상이합니다. 이 문제는 비트 조작 문제 중에서는 어려운 편인데, 제가 문제 풀이 스케줄을 짤 때 너무 초기에 배치했다는 생각이 드네요. 😓 다음 기수에서는 Number of 1 Bits처럼 좀 더 쉬운 비트 조작 문제가 먼저 나오도록 조정하도록 하겠습니다.

> 과연 이 문제에서 물어보고 싶었던 원초적인 문제의 핵심은 무엇이었을까요? :)

결국은 지원자가 `&`, `|`, `^`, `<<`, `>>`와 같은 Bitwise 연산자를 활용해서 얼마나 능숙하게 이진수를 다룰 수 있는지를 평가하기 위함입니다. 아무래도 비트 조작이 실무에서 자주 요구되는 역량은 아니기 때문에 괴리감이 느껴질 수 있습니다. 하지만 컴퓨터는 모든 데이터를 `0`과 `1`로 처리하기 때문에 CS 분야에서 기초 소양으로 여기지고 있는 것 같습니다. 실제 코딩 테스트에서는 비트 조작 문제는 온라인 시험에서나 구색 맞추기 용도로 가끔 등장하며, 대면 기술 면접에서는 다른 좋은 문제들이 워낙 많기 때문에 잘 나오지 않습니다.
